src/main/scala/com/fulcrumgenomics/math/BinomialDistribution.scala

```diff
@@ -0,0 +1,145 @@
+  * supplied to any call to [[probability()]] or [[cumulativeProbability()]].  Thus to ensure reasonable
+  * performance it is recommended to create one instance of this class and reuse it for as may
+  * calculations as is practical.
+  */
+class BinomialDistribution {
+  // This array will get expanded with more factorials any time a higher n is queried
+  private var factorials = Array(BigInt(0), BigInt(1))
+
+  // Constant values for 0 and 1 in BigDecimal to avoid making them all the time
+  private val Zero = BigDecimal(0)
+  private val One  = BigDecimal(1)
+
+  /** Get the factorial of n. Retrieves the value from a cached set of factorials. If the cache
+    * does not contain factorial(n) yet, then all factorials up to factorial(n) are computed
+    * and cached.
+    *
+    * @param n the number to compute the factorial of
+    * @return the factorial value as a BigInt
+    */
+  private def factorial(n: Int): BigInt = {
+    if (n > factorials.length - 1) {
+      // Expand the array
+      val oldLength = factorials.length
+      val tmp = new Array[BigInt](n+1)
+      System.arraycopy(factorials, 0, tmp, 0, oldLength)
+      factorials = tmp
+
+      // Fill in the new slots
+      forloop (from=oldLength, until=factorials.length) { i =>
+        factorials(i) = factorials(i-1) * BigInt(i)
+      }
+    }
+
+    factorials(n)
+  }
+
+  /** Computes the binomial coefficient, i.e. the number of combinations of k items given a total of
+    * n items, often phrased "n pick k".
+    *
+    * @param n The number of items being picked from
+    * @param k The number of items being picked (must be 0 <= k <= n)
+    * @return The number of combinations of k items from n
+   */
+  def coefficient(n: Int, k: Int): BigInt = {
+    require(k <= n, s"k must be <= n. n=$n, k=$k")
+    require(k >= 0, s"Can't compute coefficient for picking fewer than 0 values. n=$n, k=$k")
+
+    if (k == 0 || k == n) 1
+    else factorial(n) / (factorial(k) * factorial(n - k))
+  }
+
+  /**
+    * Calculates the probability of exactly `k` successes in `n` trials given `p` probability.
+    * Akin to `dbinom` in R.
+    *
+    * @param k The number of successful trials
+    * @param n The number of trials
+    * @param p The probability of the success in any one trial
+    * @return The probability of exactly k successes in n trials of p probability
+    */
+  def probability(k: Int, n: Int, p: Double): BigDecimal = probability(k, n, BigDecimal(p))
+
+  /**
+    * Calculates the probability of exactly `k` successes in `n` trials given `p` probability.
+    * Akin to `dbinom` in R.
+    *
+    * @param k The number of successful trials
+    * @param n The number of trials
+    * @param p The probability of the success in any one trial
+    * @return The probability of exactly k successes in n trials of p probability
+    */
+  def probability(k: Int, n: Int, p: BigDecimal): BigDecimal = {
```

Comment:
Is there an implicit conversion from BigDecimal to double? I'm assuming that the result is desired as a double

Reply:
I don't believe there is ... though I _want_ the result as a BigDecimal, as otherwise when summing many of them there may again be precision loss.

Reply:
Well, if there was an implicit one, it would have been OK, since cumulativeP would get the BigDecimal it needs, but a "casual" user would get a double when they need it...anyhow, it's no big deal to call doubleValue on your result....so nm.